pkg/gadgets/trace/dns/tracer/latency.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"time"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+const dnsLatencyCacheSize int = 1024
+
+// dnsQueryKey is a unique identifier for a DNS query.
+// The address is the query's source IP address (equals destination address of the response).
+// The ID comes from the DNS header.
+type dnsQueryKey struct {
+	addr [16]uint8 // Either IPv4 or IPv6.
+	id   uint16
+}
+
+// dnsLatencyCalculator calculates the latency between a query and its response.
+// It uses an LRU cache to bound memory usage.
+// All operations are thread-safe.
+type dnsLatencyCalculator struct {
+	queryCache *lru.Cache[dnsQueryKey, uint64] // This is thread-safe.
+}
+
+func newDNSLatencyCalculator() (*dnsLatencyCalculator, error) {
+	queryCache, err := lru.New[dnsQueryKey, uint64](dnsLatencyCacheSize)
+	if err != nil {
+		return nil, err
+	}
+	return &dnsLatencyCalculator{queryCache}, nil
+}
+
+func (c *dnsLatencyCalculator) storeDNSQueryTimestamp(saddr [16]uint8, id uint16, timestamp uint64) {
+	// Store the timestamp of the query so we can calculate the latency once the response arrives.
+	c.queryCache.Add(dnsQueryKey{saddr, id}, timestamp)
+}
+
+// If there is no corresponding DNS query (either never received or evicted to make space), then this returns zero.
+func (c *dnsLatencyCalculator) calculateDNSResponseLatency(daddr [16]uint8, id uint16, timestamp uint64) time.Duration {
+	key := dnsQueryKey{daddr, id}
+	reqTS, ok := c.queryCache.Get(key)
+	if !ok {
+		// Either an invalid ID or we evicted the query from the map to free space.
+		return 0
+	}
+
+	c.queryCache.Remove(key)
```

Comment:
Here the cache is removed. Since the cache is global and not per-netns, we get weird results:
```
$ ./kubectl-gadget trace dns -A -o custom-columns=timestamp,pod,id,qr,type,qtype,latency,name
TIMESTAMP                           POD                                  ID   QR TYPE      QTYPE      LATENCY              NAME                               
2023-01-16T12:16:30.882868135+01:00 normal-pod-cnl4l                     7da6 Q  OUTGOING  A                               default.svc.cluster.local.         
2023-01-16T12:16:30.882932625+01:00 coredns-565d847f94-bkdpt             7da6 Q  HOST      A                               default.svc.cluster.local.         
2023-01-16T12:16:30.882887281+01:00                                      7da6 Q  OTHERHOST A                               default.svc.cluster.local.         
2023-01-16T12:16:30.882928220+01:00                                      7da6 Q  OUTGOING  A                               default.svc.cluster.local.         
2023-01-16T12:16:30.883202854+01:00 coredns-565d847f94-bkdpt             7da6 R  OUTGOING  A          274.634µs            default.svc.cluster.local.         
2023-01-16T12:16:30.883216137+01:00                                      7da6 R  OTHERHOST A                               default.svc.cluster.local.         
2023-01-16T12:16:30.883235411+01:00                                      7da6 R  OUTGOING  A          367.276µs            default.svc.cluster.local.         
2023-01-16T12:16:30.883244999+01:00                                      7da6 R  OUTGOING  A                               default.svc.cluster.local.         
2023-01-16T12:16:30.883249613+01:00 normal-pod-cnl4l                     7da6 R  HOST      A                               default.svc.cluster.local.         
^C
Terminating...
```

When my pod `normal-pod-cnl4l` receives the dns response, the latency is not displayed because the cache has already been removed in a previous entry when the same packet has been detected in another network namespace.

Tested with:

```
$ kubectl exec -ti -n default normal-pod-cnl4l -- nslookup -type=a default.svc.cluster.local
Server:		10.96.0.10
Address:	10.96.0.10:53
```